### PR TITLE
Implement Process.warmup

### PR DESCRIPTION
**Pull request by byroot**
[[Feature #18885]](https://bugs.ruby-lang.org/issues/18885)

For now, the optimizations performed are:

  - Run a major GC
  - Compact the heap
  - Promote all surviving objects to oldgen

Other optimizations may follow.

https://github.com/ruby/ruby/pull/7346 had to be reverted because of CI failures (cc @tenderlove)

### CODE REVIEW

**Comment by byroot** [comment]
I don't get why it failed a lot on master, but passes here (one failure on Arm64 / gcc-11 YJIT, but seem unrelated)

**Comment by luke-gru** [comment]
This is a complete shot in the dark, but could it be that if the GC tests run first and compact the heap, then T_MOVED objects will exist and these aren't checked for when promoting objects?

**Comment by ko1** [comment]
Please get approval from Matz before merging, thank you!

**Comment by byroot** [comment]
@ko1 my understanding was that it was approved: https://bugs.ruby-lang.org/issues/18885#note-14, am I misunderstanding?

File: gc.c
Comment (peterzhu2118) [inline gc.c:9647]:
```suggestion
                RVALUE_AGE_SET(v, RVALUE_OLD_AGE);
                RVALUE_OLD_UNCOLLECTIBLE_SET(objspace, v);
```

Comment (byroot) [reply gc.c:9647]: Thanks that seems to have helped, there still some other weird crashes though :/

File: gc.c
Comment (peterzhu2118) [inline gc.c:9649]: I don't think you can do this. If there's a WB protected object that refers to a WB unprotected object then the WB protected object will be set as old while the WB unprotected object will not be promoted.

We should put the parent object in the remember set if it's WB unprotected. Unfortunately, I think we need to use a visitor (like marking) instead of iterating over all the objects.

Comment (byroot) [reply gc.c]: So you made me think that instead we could set them all to `RVALUE_OLD_AGE - 1` (AKA `SET_CANDIDATE`) first, and then run the major GC after.

I pushed a commit that does that.

Comment (peterzhu2118) [reply gc.c]: Yeah I think that's a cleaner solution.

File: gc.c
Comment (peterzhu2118) [inline gc.c:9687]: Is there a possibility of GC running during iteration? `gc_set_candidate_object_i` can't possibly trigger GC and it doesn't look like `rb_objspace_each_objects` can trigger GC during iteration.

File: process.c
Comment (peterzhu2118) [inline process.c:8556]: Can we add a test for this two? I think we just need to check that `GC.stat(:major_gc_count)` and `GC.stat(:compact_count)` both increased by 1.

**Review by peterzhu2118 — APPROVED** [review]
Two minor comments, otherwise looks good!

File: gc.c
Comment (casperisfine) [reply gc.c:9687]: Yeah, I think that's a leftover from the transient heap.

**Pull request merged**
